python/tests/test_tree_stats.py

```diff
@@ -44,18 +44,29 @@
 np.random.seed(5)
 
 
-def subset_combos(*args, p=0.75):
+def subset_combos(*args, p=0.5, min_tests=3):
```

Comment:
ok, so now this is accumulating a sorta random list of skipped combinations, that it will return to if there haven't been at least `min_tests` done. Also note that the random seed is now set *outside* this function, so the sequence of tests/don't test is different for each method (important).

Reply:
But this means that we'll have different tests run depending on the context - if we run one test on it's own, it'll be different to if we ran it with all other tests, or just one class. I've had this before, and it's horrible when trying to track down bugs. We should pass the random seed as a parameter. I don't think it's that important to have different test/don't test combinations for different methods, but we could come up with something that passes in a different seed which is a hash of the calling method or something like that.

Reply:
Hm. I think it's important to have different combinations for different methods, because our different tests are testing different things, so if if *all* of the combinations skip one, then we're missing that out.

Your suggestion is good, but I don't know how to get different seeds in there. We can't just put in a seed in the call to `subset_combos( )` because then every derived class will use the same seed.  Hashing the call stack, maybe, but would the call stack really be the same when running all tests and running just the one you're interested in? Seems very tricky.

Before going down this rabbit hole, let me point out that if you want to re-find a failed test, you can just set `p=1.0` in the code (eg by uncommenting line 52 below).

Reply:
That's a good point - we can always reproduce specific combos by running them all on a particular test. OK, let's not worry too much about this, sorry for the noise.

Reply:
oh, so you'd be ok with putting the seed back up to the top, rather than hashing? I'd prefer that.

Reply:
Yeah, I guess. I have a feeling it'll come back to bite us, but I don't have a better suggestion so let's just set it at the module level and run the tests non-determinisitically.